src/integration/data/vectors/upsertAndUpdate.test.ts

```diff
@@ -1,17 +1,34 @@
-import { Pinecone, Index } from '../../../index';
+/* This file tests both Upsert and Update operations. It additionally tests the RetryOnServerFailure implementation
+ on the Upsert operation. */
+
+import { Index, Pinecone } from '../../../index';
 import {
   generateRecords,
   globalNamespaceOne,
   randomIndexName,
   waitUntilReady,
 } from '../../test-helpers';
+import { RetryOnServerFailure } from '../../../utils/retries';
+import { PineconeMaxRetriesExceededError } from '../../../errors';
+import express from 'express';
+import http from 'http';
 
 // todo: add pods
 
 let pinecone: Pinecone, serverlessIndex: Index, serverlessIndexName: string;
 
+// Mock server setup for testing retries
+let mockServerlessIndex: Index;
+let callCount = 0;
+const app = express();
+app.use(express.json());
+const server: http.Server = app.listen(4000, () => {
+  console.log('Mock Pinecone server running on http://localhost:4000');
+});
+
```

Comment:
In order to test the server failures IRL, we need to set up a mock server and point Pinecone to it. I tried using `nock` originally, but with all the abstractions within the client it was a real pain in the butt to figure out the exact URL to pass at which point, so this way with `Express` is simpler and easier